### PR TITLE
[DEV-1969] Add KYC logic

### PR DESCRIPTION
**Pull request by Kolibri1990**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    private readonly storageService: DocumentStorageService,
+    private readonly kycStepRepo: KycStepRepository,
+  ) {}
+
+  @Cron(CronExpression.EVERY_DAY_AT_4AM)
+  async checkIdentSteps(): Promise<void> {
+    const allIdentInProgress = await this.kycStepRepo.findBy({
+      name: KycStepName.IDENT,
+      created: LessThan(Util.daysBefore(Config.kyc.identFailAfterDays)),
+    });
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:34]: Filter by `IN_PROGRESS`

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    for (const identStep of allIdentInProgress) {
+      identStep.fail();
+    }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:38]: Missing save

File: src/config/config.ts [hunk 3]
```diff
@@ -151,8 +151,14 @@ export class Configuration {
...
     password: process.env.KYC_PASSWORD,
     prefix: process.env.KYC_PREFIX ?? '',
+    gatewayHost: process.env.KYC_GATEWAY_HOST,
+    customerAuto: process.env.KYC_CUSTOMER_AUTO,
+    customerVideo: process.env.KYC_CUSTOMER_VIDEO,
+    apiKey: process.env.KYC_API_KEY,
+    transactionPrefix: process.env.KYC_TRANSACTION_PREFIX,
     reminderAfterDays: 2,
     failAfterDays: 7,
+    identFailAfterDays: 90,
```
Comment (davidleomay) [inline src/config/config.ts:161]: Visually separate new settings from old ones

File: src/subdomains/generic/kyc/api/kyc.controller.ts [hunk 4]
```diff
@@ -56,8 +57,13 @@ export class KycController {
...
+  @UseInterceptors(FilesInterceptor('files'))
   @ApiOkResponse()
-  async uploadDocument(@Headers(CodeHeaderName) code: string, @Param('id') id: string): Promise<KycResultDto> {
-    return this.kycService.uploadDocument(code, +id);
+  async uploadDocument(
+    @Headers(CodeHeaderName) code: string,
+    @Param('id') id: string,
+    @UploadedFiles() files: Express.Multer.File[],
+  ): Promise<KycResultDto> {
+    return this.kycService.uploadDocument(code, +id, files[0]);
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/api/kyc.controller.ts:67]: Use single file upload
```ts
@UseInterceptors(FileInterceptor('file'))
...
@UploadedFile() file: Express.Multer.File
```

File: src/subdomains/generic/kyc/entities/kyc-step.entity.ts [hunk 5]
```diff
@@ -66,7 +65,15 @@ export class KycStep extends IEntity {
     });
   }
 
+  identUrl(): string {
+    return `https://go.${Config.kyc.prefix}online-ident.ch/app/dfxauto/identifications/${this.sessionId}/identification/start`;
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/entities/kyc-step.entity.ts:70]: Move to IdentService (static method)?

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 6]
```diff
@@ -168,6 +212,17 @@ export class KycService {
...
+  getKycStepOrThrow(user: UserData, stepId: number): KycStep {
+    const kycStep = user.kycSteps.find((s) => (s.id = stepId));
+    if (!kycStep) throw new NotFoundException('Kyc step not found');
+
+    return kycStep;
+  }
+
+  async getSequenceNumber(user: UserData, stepName: KycStepName, stepType?: KycStepType): Promise<number> {
+    return user.kycSteps.filter((k) => k.name == stepName && (k.type === stepType ?? true)).length + 1;
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:224]: Move to UserData entity

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 7]
```diff
@@ -160,6 +195,15 @@ export class KycService {
...
   }
 
+  isDataComplete(user: UserData): boolean {
+    const requiredFields = ['mail', 'phone', 'firstname', 'surname', 'street', 'location', 'zip', 'country'].concat(
+      user?.accountType === AccountType.PERSONAL
+        ? []
+        : ['organizationName', 'organizationStreet', 'organizationLocation', 'organizationZip', 'organizationCountry'],
+    );
+    return requiredFields.filter((f) => !user[f]).length === 0;
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:205]: Move to UserData entity

File: src/subdomains/generic/kyc/services/integration/ident.service.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+      `Received webhook call for user ${kycStep.userData.id} (${result.identificationprocess.id}): ${result.identificationprocess.result}`,
+    );
+
+    if (IdentSucceeded(result)) {
+      kycStep.complete();
+    } else if (IdentFailed(result)) {
+      kycStep.fail();
+    } else {
+      this.logger.error(`Unknown ident result for user ${kycStep.userData.id}: ${result.identificationprocess.result}`);
+    }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/ident.service.ts:67]: Store the result to KYC step?

File: src/subdomains/generic/kyc/services/integration/ident.service.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+  // --- HELPER METHODS --- //
+
+  private async callApi<T>(url: string, method: Method = 'GET', kycStepType: KycStepType, data?: any): Promise<T> {
+    this.customer = kycStepType == KycStepType.AUTO ? Config.kyc.customerAuto : Config.kyc.customerVideo;
+    return this.request<T>(url, method, data).catch((e: HttpError) => {
+      this.logger.verbose(`Error during intrum request ${method} ${url}: ${e.response?.status} ${e.response?.data}`);
+      throw new ServiceUnavailableException({ status: e.response?.status, data: e.response?.data });
+    });
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/ident.service.ts:84]: Have `customer` as parameter, not class variable (might give conflicts).

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    const allIdentInProgress = await this.kycStepRepo.findBy({
+      name: KycStepName.IDENT,
+      status: KycStepStatus.IN_PROGRESS,
+      created: LessThan(Util.daysBefore(Config.kyc.identFailAfterDays)),
+    });
+
+    for (const identStep of allIdentInProgress) {
+      identStep.fail();
+      await this.kycStepRepo.save(identStep);
+    }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:39]: Call fail via `UserData.failStep`

File: src/subdomains/generic/kyc/services/integration/ident.service.ts (hunk 8 above)
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/ident.service.ts:67]: Call complete/fail via UserData

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    const user = await this.getUserOrThrow(kycHash);
+    return KycInfoMapper.toDto(user);
+  }
+
+  async updatePersonalData(kycHash: string, stepId: number): Promise<KycResultDto> {
+    const user = await this.getUserOrThrow(kycHash);
+    const kycStep = this.getKycStepOrThrow(user, stepId);
+
+    if (kycStep.userData.isDataComplete()) kycStep.complete();
+    // TODO: update user data and complete step (if data complete), might be mail or personal data step
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:52]: `user.isDataComplete()`

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+      document.mimetype as KycContentType,
+      {
+        document: document.mimetype.toString(),
+        creationTime: new Date().toISOString(),
+        fileName: document.filename,
+      },
+    );
+
+    if (url) kycStep.complete();
+
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:95]: Complete via UserData.

Comment (davidleomay) [reply src/subdomains/generic/kyc/services/kyc.service.ts:95]: Store url as result?

File: src/subdomains/generic/kyc/services/integration/ident.service.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    const userData = {
+      birthday: user.birthday,
+      birthplace: user.location,
+      email: user.mail,
+      firstname: user.firstname,
+      lastname: user.surname,
+      mobilephone: user.phone,
+      nationality: user.nationality,
+      zipcode: user.zip,
+    };
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/ident.service.ts:30]: Do we need to send this data? 🤔

Comment (Kolibri1990) [reply src/subdomains/generic/kyc/services/integration/ident.service.ts:30]: It's not mandatory. I will remove.

File: src/subdomains/generic/kyc/services/integration/ident.service.ts [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+  constructor(private readonly http: HttpService) {}
+
+  async initiateIdent(kycStepType: KycStepType, transactionId: string): Promise<{ id: string }> {
+    return this.callApi<{ id: string }>(`identifications/${transactionId}/start`, kycStepType, 'POST');
+  }
+
+  static identUrl(kycStep: KycStep): string {
+    return `https://go.online-ident.ch/app/dfxauto/identifications/${kycStep.sessionId}/identification/start`;
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/ident.service.ts:23]: `dfxauto` should be replaced by new customer (auto/video)?

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+    let user = kycStep.userData;
+
+    this.logger.info(`Received ident webhook call for user ${user.id} (${sessionId}): ${sessionStatus}`);
+
+    if (IdentSucceeded(dto)) {
+      user = user.completeStep(kycStep, result);
+    } else if (IdentPending(dto)) {
+      // TODO
+      // user = user.stepInReview(kycStep);
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:109]: Add `InReview` status to KYC step (+ store result)

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+      case KycStepName.PERSONAL_DATA:
+        if (user.isDataComplete) kycStep.complete();
+        break;
+
+      case KycStepName.IDENT:
+        // TODO: verify 2FA
+
+        const transactionId = `${Config.kyc.transactionPrefix}-${kycStep.id}-${nextSequenceNumber}`;
+        const { id } = await this.identService.initiateIdent(stepType, transactionId);
+        kycStep.sessionId = id;
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:227]: Doesn't work, `kycStep.id` is undefined (not saved yet) 🤔

File: src/subdomains/generic/kyc/services/integration/ident.service.ts [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+export class IdentService {
+  private readonly logger = new DfxLogger(IdentService);
+
+  private readonly baseUrl = `${Config.kyc.gatewayHost}/api/v1`;
+
+  constructor(private readonly http: HttpService) {}
+
+  async initiateIdent(kycStepType: KycStepType, transactionId: string): Promise<{ id: string }> {
+    return this.callApi<{ id: string }>(`identifications/${transactionId}/start`, kycStepType, 'POST');
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/integration/ident.service.ts:19]: I think this doesn't work anymore 🤔

**Review by davidleomay — APPROVED** [review]

**Pull request merged**